lib/graph_gateway.go

```diff
@@ -0,0 +1,594 @@
+	exch := &handoffExchange{
+		startingExchange: carFetchingExch,
+		followupExchange: &blockFetcherExchWrapper{api.blockFetcher},
+		handoffCh:        doneWithFetcher,
+	}
+
+	api.lk.Lock()
+	api.notifiers[exch] = struct{}{}
+	api.lk.Unlock()
+
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				fmt.Println("Recovered fetcher error", r)
+			}
+		}()
+		err := api.fetcher.Fetch(ctx, path, func(resource string, reader io.Reader) error {
+			cr, err := car.NewCarReader(reader)
+			if err != nil {
+				return err
+			}
+			for {
+				blk, err := cr.Next()
+				if err != nil {
+					if errors.Is(err, io.EOF) {
+						return nil
+					}
+					return err
+				}
+				if err := bstore.Put(ctx, blk); err != nil {
+					return err
+				}
+				api.notifyAllOngoingRequests(ctx, blk)
+			}
+		})
+		if err != nil {
+			graphLog.Error(err)
+		}
+		if err := carFetchingExch.Close(); err != nil {
+			graphLog.Error(err)
+		}
+		doneWithFetcher <- struct{}{}
+		close(doneWithFetcher)
+	}()
+
+	bserv := blockservice.New(bstore, exch)
+	blkgw, err := gateway.NewBlocksGateway(bserv)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return blkgw, func() {
+		api.lk.Lock()
+		delete(api.notifiers, exch)
+		api.lk.Unlock()
+	}, nil
+}
+
+func (api *GraphGateway) notifyAllOngoingRequests(ctx context.Context, blks ...blocks.Block) {
+	api.lk.RLock()
+	for n := range api.notifiers {
+		err := n.NotifyNewBlocks(ctx, blks...)
+		if err != nil {
+			graphLog.Error(fmt.Errorf("notifyAllOngoingRequests failed: %w", err))
+		}
+	}
+	api.lk.RUnlock()
+}
+
+type fileCloseWrapper struct {
+	files.File
+	closeFn func()
+}
+
+func (w *fileCloseWrapper) Close() error {
+	w.closeFn()
+	return w.File.Close()
+}
+
+type dirCloseWrapper struct {
+	files.Directory
+	closeFn func()
+}
+
+func (w *dirCloseWrapper) Close() error {
+	w.closeFn()
+	return w.Directory.Close()
+}
+
+func wrapNodeWithClose[T files.Node](node T, closeFn func()) (T, error) {
+	var genericNode files.Node = node
+	switch n := genericNode.(type) {
+	case *files.Symlink:
+		closeFn()
+		return node, nil
+	case files.File:
+		var f files.File = &fileCloseWrapper{n, closeFn}
+		return f.(T), nil
+	case files.Directory:
+		var d files.Directory = &dirCloseWrapper{n, closeFn}
+		return d.(T), nil
+	default:
+		closeFn()
+		var zeroType T
+		return zeroType, fmt.Errorf("unsupported node type")
+	}
+}
+
+func (api *GraphGateway) Get(ctx context.Context, path gateway.ImmutablePath) (gateway.ContentPathMetadata, *gateway.GetResponse, error) {
+	blkgw, closeFn, err := api.loadRequestIntoSharedBlockstoreAndBlocksGateway(ctx, path.String()+"?format=car&depth=1")
+	if err != nil {
+		return gateway.ContentPathMetadata{}, nil, err
+	}
+	md, gr, err := blkgw.Get(ctx, path)
+	if err != nil {
+		return gateway.ContentPathMetadata{}, nil, err
+	}
+	//TODO: interfaces here aren't good enough so we're getting around the problem this way
+	runtime.SetFinalizer(gr, func(_ *gateway.GetResponse) { closeFn() })
```

Comment:
@lidel some of this is due to the implementation concerns, but others are from the gateway API. Will raise a PR in boxo to discuss